Processing version/QTree/TheQuadTree.pde

```diff
@@ -0,0 +1,97 @@
+class QuadTree {
+  Rectangle boundary;
+  int capacity;
+  ArrayList<Point> points;
+  Boolean isDivided;
+
+  QuadTree northEast, northWest, southEast, southWest;
+
+  QuadTree(Rectangle boundary, int n) {
+    this.boundary = new Rectangle(boundary);
+    this.capacity = n;
+    this.points = new ArrayList<Point>();
+    this.isDivided = false;
+  }
+
+  void subdivide() {
+    float x = this.boundary.x;
+    float y = this.boundary.y;
+    float w = this.boundary.w * 0.5;
+    float h = this.boundary.h * 0.5;
+
+    Rectangle ne = new Rectangle(x + w, y - h, w, h);
+    northEast = new QuadTree(ne, this.capacity);
+    Rectangle nw = new Rectangle(x - w, y - h, w, h);
+    northWest = new QuadTree(nw, this.capacity);
+    Rectangle se = new Rectangle(x + w, y + h, w, h);
+    southEast = new QuadTree(se, this.capacity);
+    Rectangle sw = new Rectangle(x - w, y + h, w, h);
+    southWest = new QuadTree(sw, this.capacity);
+    isDivided = true;
+  }
+
+  Boolean insert(Point point) {
+    if (!this.boundary.contains(point)) {
+      return false;
+    }
+
+    if (points.size() < this.capacity) {
+      points.add(point);
+      return true;
+    } else {
+      if (!isDivided) {
+        subdivide();
+      }
+      if (northEast.insert(point)) {
+        return true;
+      } else if (this.northWest.insert(point)) {
+        return true;
+      } else if (this.southEast.insert(point)) {
+        return true;
+      } else if (this.southWest.insert(point)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  ArrayList<Point> query(Rectangle range, ArrayList<Point> found) {
```

Comment:
Should this be List rather than ArrayList?

Reply:
`ArrayList` is good b/c it's what I use as standard catch-all flexible array in my tutorials and examples!